shelley/chain-and-ledger/executable-spec/test/Generator/Utxo/QuickCheck.hs

```diff
@@ -130,13 +136,32 @@ genTx (LedgerEnv slot _ pparams _) (utxoSt@(UTxOState utxo _ _ _), dpState) keys
     -- deterministically for each script. Varying the script is possible though.
     let keysLists = map getKeyCombination $ Map.elems multiSig
         msigSignatures = foldl Set.union Set.empty $ map Set.fromList keysLists
-        !wits = makeWitnessesVKey txBody (spendWitnesses ++ certWitnesses ++ wdrlWitnesses ++ updateWitnesses)
-                `Set.union` makeGenWitnessesVKey txBody genesisWitnesses
-                `Set.union` makeWitnessesFromScriptKeys txBody keyHashMap msigSignatures
+        wits = makeWitnessesVKey txBody (spendWitnesses ++ certWitnesses ++ wdrlWitnesses ++ updateWitnesses)
+          `Set.union` makeGenWitnessesVKey txBody genesisWitnesses
+          `Set.union` makeWitnessesFromScriptKeys txBody keyHashMap msigSignatures
 
     let metadata = Nothing -- TODO generate metadata
-    -- discard if balance is negative, i.e., deposits exceed spending balance
-    pure (Tx txBody wits multiSig metadata)
+
+    -- calculate real fees of witnesses transaction
+    let minimalFees = minfee pparams (Tx txBody wits multiSig metadata)
+
+    -- discard generated transaction if the balance cannot cover the fees
+    if minimalFees > balance_
+      then D.trace (  "discarded bc. of real fees, minimal: "
+                   ++ show minimalFees ++ " Output balance: "
+                   ++ show balance_
+                   ) QC.discard
+      else do
+
+      -- update model transaction with real fees and outputs
+      let (fees', outputs') = calcFeeAndOutputs balance_ recipientAddrs minimalFees
+          txBody' = txBody { _txfee = fees'
+                           , _outputs = outputs' }
+          wits' = makeWitnessesVKey txBody' (spendWitnesses ++ certWitnesses ++ wdrlWitnesses ++ updateWitnesses)
+            `Set.union` makeGenWitnessesVKey txBody' genesisWitnesses
+            `Set.union` makeWitnessesFromScriptKeys txBody' keyHashMap msigSignatures
```

Comment:
let's abstract this since we're re-using it in genTx

Reply:
done